src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,550 @@
+                new DataflowLinkOptions { PropagateCompletion = true },
+                cancellationToken: ProjectAsynchronousTasksService.UnloadCancellationToken);
+        }
+
+        /// <summary>
+        /// Preprocess gets called as each data flow block updates and its job is to take the input from those blocks and do whatever work needed
+        /// so that ApplyAsync has all of the info it needs to do its job.
+        /// </summary>
+        protected override Task<DesignTimeInputsDelta> PreprocessAsync(IProjectVersionedValue<InputTuple> input, DesignTimeInputsDelta previousOutput)
+        {
+            IProjectChangeDescription compileChanges = input.Value.Item1.ProjectChanges[Compile.SchemaName];
+            IProjectChangeDescription configChanges = input.Value.Item2.ProjectChanges[ConfigurationGeneral.SchemaName];
+
+            ImmutableArray<string>.Builder addedDesignTimeInputs = ImmutableArray.CreateBuilder<string>();
+            ImmutableArray<string>.Builder removedDesignTimeInputs = ImmutableArray.CreateBuilder<string>();
+            ImmutableArray<string>.Builder addedDesignTimeSharedInputs = ImmutableArray.CreateBuilder<string>();
+            ImmutableArray<string>.Builder removedDesignTimeSharedInputs = ImmutableArray.CreateBuilder<string>();
+
+            foreach (string item in compileChanges.Difference.AddedItems)
+            {
+                PreprocessAddItem(item);
+            }
+
+            foreach (string item in compileChanges.Difference.RemovedItems)
+            {
+                PreprocessRemoveItem(item);
+            }
+
+            foreach ((string oldName, string newName) in compileChanges.Difference.RenamedItems)
+            {
+                // A rename is just an add and a remove
+                PreprocessAddItem(newName);
+                PreprocessRemoveItem(oldName);
+            }
+
+            foreach (string item in compileChanges.Difference.ChangedItems)
+            {
+                (bool wasDesignTime, bool wasDesignTimeShared) = GetDesignTimePropsForItem(compileChanges.Before.Items[item]);
+                (bool designTime, bool designTimeShared) = GetDesignTimePropsForItem(compileChanges.After.Items[item]);
+
+                if (!wasDesignTime && designTime)
+                {
+                    addedDesignTimeInputs.Add(item);
+                }
+                else if (wasDesignTime && !designTime)
+                {
+                    removedDesignTimeInputs.Add(item);
+                }
+                if (!wasDesignTimeShared && designTimeShared)
+                {
+                    addedDesignTimeSharedInputs.Add(item);
+                }
+                else if (wasDesignTimeShared && !designTimeShared)
+                {
+                    removedDesignTimeSharedInputs.Add(item);
+                }
+            }
+
+            bool namespaceChanged = configChanges.Difference.ChangedProperties.Contains(ConfigurationGeneral.RootNamespaceProperty);
```

Comment:
Does this work for VB? They have a slightly different representation here that I cannot remember.

Reply:
VB handles RootNamespace differently in the compiler, which is what https://github.com/dotnet/roslyn/pull/33653 fixed, but I didn't run into any problems here in the testing that led to that PR.